### PR TITLE
mon/MonClient: handle ms_handle_fast_authentication return

### PR DESCRIPTION
**Pull request by batrick**
Fixes: https://tracker.ceph.com/issues/62382

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by batrick** [comment]
I think to test this would require having caps which parse correctly for the mon's `AuthMonitor` but are too new to parse for an older OSD/mgr/mds. Would you agree @idryomov @rzarzynski ?

**Comment by idryomov** [comment]
> I think to test this would require having caps which parse correctly for the mon's `AuthMonitor` but are too new to parse for an older OSD/mgr/mds.

Most likely, given that the only thing these handlers do is parse the caps string.

File: src/mon/MonClient.cc
Comment (batrick) [reply src/mon/MonClient.cc:1644]: thanks, fixed

File: src/msg/Dispatcher.h
Comment (batrick) [reply src/msg/Dispatcher.h]: thanks, fixed

File: src/mon/Monitor.cc
Comment (batrick) [reply src/mon/Monitor.cc:6583]: thanks, fixed

**Comment by batrick** [comment]
> > I think to test this would require having caps which parse correctly for the mon's `AuthMonitor` but are too new to parse for an older OSD/mgr/mds.
> 
> Most likely, given that the only thing these handlers do is parse the caps string.

Okay, I'll have to think of an easy way to test this. Alternatively we can just skip tests as risk is low, IMO.

File: src/msg/Dispatcher.h
Comment (batrick) [inline src/msg/Dispatcher.h:219]: @idryomov actually this should probably be `true`, yes? Otherwise every dispatcher that does not implement this method will fail here. I'll make this change.

**Comment by batrick** [comment]
> > I think to test this would require having caps which parse correctly for the mon's `AuthMonitor` but are too new to parse for an older OSD/mgr/mds.
> 
> Most likely, given that the only thing these handlers do is parse the caps string.

Given how brittle the test would be, I think we'll just take this as-is. Please review at your convenience @idryomov .

File: src/mon/Monitor.cc
Comment (idryomov) [reply src/mon/Monitor.cc:6600]: Additionally, there is a bunch of dispatchers both in production (`OSD::HeartbeatDispatch`) and test code which need to be updated to the new signature.  Please do at least a minimal build before asking for review ;)

File: src/msg/Dispatcher.h
Comment (idryomov) [reply src/msg/Dispatcher.h]: I think it's dangerous to provide a default "yup, authenticated" implementation.  I would rather make this a pure virtual.

**Review by idryomov — COMMENTED** [review]
I see two `ms_handle_fast_authentication()` calls in `Monitor::handle_auth_request()` where the return value is not checked.  Is that intentional?

**Comment by idryomov** [comment]
> Given how brittle the test would be, I think we'll just take this as-is.

This is up to @rzarzynski.

File: src/mon/Monitor.cc
Comment (batrick) [reply src/mon/Monitor.cc:6600]: Thanks and apologies. It is fixed now.

**Comment by batrick** [comment]
> I see two ms_handle_fast_authentication() calls in Monitor::handle_auth_request() where the return value is not checked. Is that intentional?

Taking a look at this now.

File: src/msg/Dispatcher.h
Comment (rzarzynski) [inline src/msg/Dispatcher.h:219]: Any reason for not going pure virtual?

Comment (batrick) [reply src/msg/Dispatcher.h:219]: Dispatchers that don't handle authentication would need to have the method which seemed unnecessarily obtrusive.

Comment (rzarzynski) [reply src/msg/Dispatcher.h:219]: I share the @idryomov's concern.

How about introducing an intermediary stage (with the default implementation) in the inheritance hierarchy to address those auth-uninterested dispatchers explicitly?

```
$ grep -r ms_handle_fast_authentication ./
./mon/MonClient.cc:      handle_authentication_dispatcher->ms_handle_fast_authentication(con);
./mon/MonClient.cc:    handle_authentication_dispatcher->ms_handle_fast_authentication(con);
./mon/Monitor.h:  int ms_handle_fast_authentication(Connection *con) override;
./mon/AuthMonitor.cc:	  mon.ms_handle_fast_authentication(s->con.get()) > 0) {
./mon/Monitor.cc:      ms_handle_fast_authentication(con);
./mon/Monitor.cc:    ms_handle_fast_authentication(con);
./mon/Monitor.cc:int Monitor::ms_handle_fast_authentication(Connection *con)
./msg/Dispatcher.h:  virtual int ms_handle_fast_authentication(Connection *con) {
./mds/MDSDaemon.h:  int ms_handle_fast_authentication(Connection *con) override;
./mds/MDSDaemon.cc:int MDSDaemon::ms_handle_fast_authentication(Connection *con)
./mgr/DaemonServer.h:  int ms_handle_fast_authentication(Connection *con) override;
./mgr/DaemonServer.cc:int DaemonServer::ms_handle_fast_authentication(Connection *con)
./osd/OSD.h:    int ms_handle_fast_authentication(Connection *con) override {
./osd/OSD.h:  int ms_handle_fast_authentication(Connection *con) override;
./osd/OSD.cc:int OSD::ms_handle_fast_authentication(Connection *con)
./test/fio/fio_ceph_messenger.cc:  int ms_handle_fast_authentication(Connection *con) override {
./test/msgr/perf_msgr_client.cc:    int ms_handle_fast_authentication(Connection *con) override {
./test/msgr/test_msgr.cc:  int ms_handle_fast_authentication(Connection *con) override {
./test/msgr/test_msgr.cc:  int ms_handle_fast_authentication(Connection *con) override {
./test/msgr/test_msgr.cc:  int ms_handle_fast_authentication(Connection *con) override {
./test/msgr/perf_msgr_server.cc:  int ms_handle_fast_authentication(Connection *con) override {
```

```
$ grep -r public\ Dispatcher ./
./mon/Monitor.h:class Monitor : public Dispatcher,
./mon/MonClient.h:struct MonClientPinger : public Dispatcher,
./mon/MonClient.h:class MonClient : public Dispatcher,
./neorados/RADOSImpl.h:class RADOS : public Dispatcher
./tools/cephfs_mirror/ClusterWatcher.h:class ClusterWatcher : public Dispatcher {
./tools/cephfs/MDSUtility.h:class MDSUtility : public Dispatcher {
./librados/RadosClient.h:class librados::RadosClient : public Dispatcher,
./librbd/io/ImageDispatcher.h:class ImageDispatcher : public Dispatcher<ImageCtxT, ImageDispatcherInterface> {
./librbd/io/ObjectDispatcherInterface.h:  : public DispatcherInterface<ObjectDispatchInterface> {
./librbd/io/ObjectDispatcher.h:  : public Dispatcher<ImageCtxT, ObjectDispatcherInterface> {
./librbd/io/ImageDispatcherInterface.h:  : public DispatcherInterface<ImageDispatchInterface> {
./osdc/Objecter.h:class Objecter : public md_config_obs_t, public Dispatcher {
./mds/MDSDaemon.h:class MDSDaemon : public Dispatcher {
./mds/Beacon.h:class Beacon : public Dispatcher
./mds/MetricsHandler.h:class MetricsHandler : public Dispatcher {
./mds/MetricAggregator.h:class MetricAggregator : public Dispatcher {
./mgr/MgrStandby.h:class MgrStandby : public Dispatcher,
./mgr/DaemonServer.h:class DaemonServer : public Dispatcher, public md_config_obs_t
./mgr/MgrClient.h:class MgrClient : public Dispatcher
./osd/OSD.h:class OSD : public Dispatcher,
./osd/OSD.h:  struct HeartbeatDispatcher : public Dispatcher {
./test/crimson/test_messenger.cc:class FailoverSuite : public Dispatcher {
./test/crimson/test_messenger.cc:class FailoverTest : public Dispatcher {
./test/crimson/test_messenger.cc:class FailoverSuitePeer : public Dispatcher {
./test/crimson/test_messenger.cc:class FailoverTestPeer : public Dispatcher {
./test/crimson/test_messenger_peer.cc:class FailoverSuitePeer : public Dispatcher {
./test/crimson/test_messenger_peer.cc:class FailoverTestPeer : public Dispatcher {
./test/fio/fio_ceph_messenger.cc:class FioDispatcher : public Dispatcher {
./test/mon/test_mon_workloadgen.cc:class TestStub : public Dispatcher
./test/mon/test-mon-msg.cc:class MonClientHelper : public Dispatcher
./test/direct_messenger/test_direct_messenger.cc:class MockDispatcher : public Dispatcher {
./test/testmsgr.cc:class Admin : public Dispatcher {
./test/msgr/perf_msgr_client.cc:  class ClientDispatcher : public Dispatcher {
./test/msgr/test_msgr.cc:class FakeDispatcher : public Dispatcher {
./test/msgr/test_msgr.cc:class SyntheticDispatcher : public Dispatcher {
./test/msgr/test_msgr.cc:class MarkdownDispatcher : public Dispatcher {
./test/msgr/perf_msgr_server.cc:class ServerDispatcher : public Dispatcher {
./client/Client.h:class Client : public Dispatcher, public md_config_obs_t {
```

For many of them the burden could as low as running `sed`.

Comment (batrick) [reply src/msg/Dispatcher.h]: I think my concern in

https://github.com/ceph/ceph/pull/52939#discussion_r1367077257

was not actually an issue. So I've changed the default return to false. That should render this conversation moot.

Comment (batrick) [reply src/msg/Dispatcher.h]: Well, I do need to adjust some daemons. Hmm.

**Comment by batrick** [comment]
@rzarzynski please have another look.

**Comment by batrick** [comment]
(trivial rebase)

**Comment by batrick** [comment]
@rzarzynski @idryomov review refresh please.

**Comment by batrick** [comment]
trivial rebase

**Comment by batrick** [comment]
jenkins test windows

**Comment by batrick** [comment]
This PR is under test in [https://tracker.ceph.com/issues/66433](https://tracker.ceph.com/issues/66433).

**Comment by batrick** [comment]
This PR is under test in [https://tracker.ceph.com/issues/66462](https://tracker.ceph.com/issues/66462).

**Comment by batrick** [comment]
This PR is under test in [https://tracker.ceph.com/issues/66609](https://tracker.ceph.com/issues/66609).

**Comment by batrick** [comment]
@idryomov @rzarzynski this is ready for merge; just blocked on your review please.

File: src/mgr/DaemonServer.h
Comment (rzarzynski) [inline src/mgr/DaemonServer.h:275]: Proposal: how about marking this as `[[nodiscard]]` to rise a warning if the return goes unchecked?

**Review by rzarzynski — APPROVED** [review]

File: src/mgr/DaemonServer.h
Comment (batrick) [reply src/mgr/DaemonServer.h:275]: Sure!

**Comment by batrick** [comment]
jenkins test make check arm64

**Comment by batrick** [comment]
jenkins test make check

**Comment by batrick** [comment]
jenkins test make check

**Comment by batrick** [comment]
jenkins test api

**Comment by batrick** [comment]
jenkins test make check arm64

**Comment by batrick** [comment]
This PR is under test in [https://tracker.ceph.com/issues/66822](https://tracker.ceph.com/issues/66822).

**Comment by batrick** [comment]
jenkins test make check arm64

**Comment by batrick** [comment]
jenkins test make check arm64

**Comment by batrick** [comment]
This PR is under test in [https://tracker.ceph.com/issues/67214](https://tracker.ceph.com/issues/67214).

**Pull request merged**
